arrow-core-data/src/main/kotlin/arrow/core/computations/nullable.kt

```diff
@@ -1,18 +1,17 @@
 package arrow.core.computations
 
-import arrow.continuations.generic.DelimContScope
-import arrow.continuations.generic.DelimitedScope
-import arrow.core.computations.suspended.BindSyntax
+import arrow.continuations.Effect
+import arrow.continuations.Reset
+
+fun interface NullableEffect<A> : Effect<A?> {
+  suspend operator fun <B> B?.invoke(): B = this ?: control().shift(null)
+}
 
 @Suppress("ClassName")
-object nullable {
-  operator fun <A> invoke(func: suspend BindSyntax.() -> A?): A? =
-    DelimContScope.reset { NullableBindSyntax(this).func() }
+object nullable
 
-  private class NullableBindSyntax<R>(
-    scope: DelimitedScope<R?>
-  ) : BindSyntax, DelimitedScope<R?> by scope {
-    override suspend fun <A> A?.invoke(): A =
-      this ?: shift { null }
-  }
-}
+fun <A> nullable.eager(func: suspend NullableEffect<A>.() -> A?): A? =
```

Comment:
Allows for foreign extension. We need to define this seperately, and then it's safe.

```
@RestrictsSuspension
fun interface EagerNullableEffect<A> : Effect<A?> {
  suspend operator fun <B> B?.invoke(): B = this ?: control().shift(null)
}
```

```suggestion
fun <A> nullable.eager(func: suspend EagerNullableEffect<A>.() -> A?): A? =
```